core/unit_test/incremental/Test02c_MemorySpace.hpp

```diff
@@ -0,0 +1,140 @@
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:2
+
+// Deep copy unit tests
+// We allocate and initialize an array on the Host.
+// Copy the array to the Device and copy back the same from Device To Host in a
+// different array. Check if the original array values match with the values in
+// the array copied back from the Device.
+
+namespace Test {
+
+using value_type       = double;
+const int num_elements = 10;
+const value_type value = 0.5;
+
+template <class ExecSpace>
+struct TestIncrMemorySpace_deepcopy {
+  // Memory space type for Device and Host data
+  using d_memspace_type = typename ExecSpace::memory_space;
+  using h_memspace_type = Kokkos::HostSpace;
+
+  // Routine to check correctness.
+  // Checks whether the original host array and the array copied/received from
+  // the Device are exactly the same.
+  void correctness_check(value_type *hostData, value_type *hostData_recv) {
+    // Check if the size of the originally allocated buffer and the buffer
+    // received from the Device are of equal length.
+    EXPECT_EQ(sizeof(hostData), sizeof(hostData_recv))
+        << "Original hostData and the copied back hostData_recv are of unequal "
+           "length";
+
+    // Check the element-wise equality of both the buffers.
+    for (int i = 0; i < num_elements; ++i) {
+      ASSERT_EQ(hostData[i], hostData_recv[i])
+          << "hostData and hostData_recv differ at index" << i;
+    }
+  }
+
+  // Routine to allocate memory in a specific memory space.
+  template <class MemSpace>
+  value_type *allocate_mem(int N) {
+    return (static_cast<value_type *>(
+        Kokkos::kokkos_malloc<MemSpace>("deviceData", N * sizeof(value_type))));
```

Comment:
Technically, you are not always allocating device data using this function.

Reply:
Do you mean there is a data allocation that is not happening via "allocate_mem" routine?

Reply:
No, I am saying that the label "deviceData" might be a bit misleading since you also use this function to allocate memory in host space.

Reply:
Ah yes. Fixed it.